scripts/generate/templates/StorageSlot.js

```diff
@@ -39,40 +30,179 @@ pragma solidity ^0.8.20;
  */
 `;
 
-const struct = type => `\
-struct ${type.struct} {
-  ${type.type} value;
+const tooling = `\
+/**
+ * @dev Derive an ERC-1967 slot from a string (path).
+ */
+function erc1967slot(string memory path) internal pure returns (bytes32 slot) {
+  /// @solidity memory-safe-assembly
+  assembly {
+      slot := sub(keccak256(add(path, 0x20), mload(path)), 1)
+  }
+}
+
+/**
+ * @dev Derive an ERC-7201 slot from a string (path).
+ */
+function erc7201slot(string memory path) internal pure returns (bytes32 slot) {
+  /// @solidity memory-safe-assembly
+  assembly {
+      mstore(0x00, sub(keccak256(add(path, 0x20), mload(path)), 1))
+      slot := and(keccak256(0x00, 0x20), not(0xff))
+  }
+}
+
+/**
+ * @dev Add an offset to a slot to get the n-th element of a structure or an array.
+ */
+function offset(bytes32 slot, uint256 pos) internal pure returns (bytes32 result) {
+  unchecked {
+      return bytes32(uint256(slot) + pos);
+  }
+}
+
+/**
+ * @dev Derive the location of the first element in an array from the slot where the length is stored.
+ *
+ * See: https://docs.soliditylang.org/en/v0.8.20/internals/layout_in_storage.html#mappings-and-dynamic-arrays.
```

Comment:
I think we should reference these as a link (same applies to others):

```suggestion
 * See https://docs.soliditylang.org/en/v0.8.20/internals/layout_in_storage.html#mappings-and-dynamic-arrays[Solidity docs for mappings and dynamic arrays.]
```